### PR TITLE
Detail the new login migration impacts

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 9.x
| Description?  | Detail the new login migration impacts
| Fixed ticket? | ~

### CODE REVIEW

File: modules/core-updates/9.0.md
Comment (Hlavtox) [inline modules/core-updates/9.0.md]: @jolelievre Jo, will this $context->cookie still be accessible or not? Because some things depend on it, For example - https://github.com/PrestaShop/ps_googleanalytics/blob/618ade8f9a424ad31968f6367ddf3e0be8c5f879/ps_googleanalytics.php#L131

If not, could you note some other way to store temporary data between sessions? ($session?)

File: modules/core-updates/9.0.md
Comment (PrestaEdit) [inline modules/core-updates/9.0.md]:
```suggestion
The logic of the login, however, is not handled by PrestaShop internal code so there is no handler that interprets the submitted data (and that's why there is so few new hooks), instead we rely on <a href="https://symfony.com/doc/current/security.html#form-login" target="_blank">Symfony form login authentication system</a>, so it's Symfony internal system that interprets the data submitted, validates it and authenticate the employee to the back-office.
```

File: modules/core-updates/9.0.md
Comment (jolelievre) [reply modules/core-updates/9.0.md]: Hmm actually you can still write data in the cookie, it's just that the core no longer depends on it So for example you can't perform a BO login by manipulating the cookie anymore
But persisted data is still present in the cookie so it should be compatible for the modules, but you're right I'm gonna detail this a bit more

And regarding new possibilities to persist data, you can save attributes in the Symfony token That's what we did to handle:
- shop context
- employee session
- IP address checking

So it's worth detailing this as well I guess.

**Review by matthieu-rolland — APPROVED** [review]

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
The back office login page has been migrated to Symfony. Along with this change, the authorization system in the back office is now also based on Symfony, which implies several things:
```

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
- we no longer depend on the legacy `Context::$cookie`, the session is kept on the server side, and very few data are kept on the browser side, for retro-compatibility we still populate the legacy cookie so that you can **read** it, but modifying/writing it will likely no longer have the effect you would expect (if any)
```

as @Hlavtox mentioned, it requires some clarification because it is probably not true for most of the scenarios cookies have been used by community developers

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
- the `PrestaShopBundle\Security\Admin\EmployeeProvider` and `PrestaShopBundle\Service\DataProvider\UserProvider` now return a `PrestaShopBundle\Entity\Employee` instance, their responsibility is to return a `Symfony\Component\Security\Core\User\UserInterface` anyway (and they still do) but in case you depended on the child class know that `PrestaShopBundle\Security\Admin\Employee` no longer exist
```

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
- if you need to get the logged-in user, you can use the Symfony `Symfony\Bundle\SecurityBundle\Security` service, but we recommend you use the `EmployeeContext` the internal PrestaShop service (see explanation about [new Contexts]({{< relref "9/modules/core-updates/9.0.md#symfony-context-refacto" >}}))
```

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
The login logic is not handled by PrestaShop's internal code, so there is no handler that interprets the submitted data (and that's why there are so few new hooks). Instead, we rely on <a href="https://symfony.com/doc/current/security.html#form-login" target="_blank">Symfony form login authentication system</a>, so it's Symfony's internal system that interprets the data submitted, validates it and authenticate the employee to the back office.
```

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
If you need to interfere with the authentication process, you should now rely on the <a href="https://symfony.com/doc/current/security.html#authentication-events" target="_blank">Authentication events</a> described in their documentation, you can also inspire yourself from our <a href="https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/EventListener/Admin/EmployeeSessionSubscriber.php" target="_blank">own internal subscriber</a>.
```

**Review by kpodemski — COMMENTED** [review]
A few comments.

Any particular reason not to use Markdown URLs? Is it because of the `target="_blank"`?

**Comment by jolelievre** [comment]
> A few comments.
> 
> Any particular reason not to use Markdown URLs? Is it because of the `target="_blank"`?

Yep, I couldn't find a way to have target blank with the markdown syntax, and for external documentation I thought it was better to open them in another tab.

File: modules/core-updates/9.0.md
Comment (jolelievre) [reply modules/core-updates/9.0.md]: Yeah I modified this, let me know after I push my changes it it's clearer

File: modules/core-updates/9.0.md
Comment (jolelievre) [reply modules/core-updates/9.0.md]: I prefered the word `integrate`, interfere makes it feel like you wanna prevent the initial behaviour more than add something to it

**Comment by jolelievre** [comment]
@Hlavtox @kpodemski do you think it's clearer now?

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
- we no longer depend on the legacy `Context::$cookie`, the session is kept on the server side, and very few data are kept on the browser side. For retro-compatibility, we still populate the legacy cookie so that you can **read** it. While you can work with `Context::$cookie` and your custom data, making changes to the data used for PrestaShop authorization may no longer have the desired effect or may result in instability, so it's not recommended.
```

A little rephrase.

**Comment by Hlavtox** [comment]
@jolelievre Yes, all clear now! :-)

-------------

By the way, is the recommended way to store some temporary data the symfony session?
If yes, we can add a new section in the docs and link it. Something like:

In the past, if you needed to store lightweight temporary data for the session, you could use the Prestashop cookie for this purpose. However, with the recent migration of login system to symfony, this is slowly getting obsolete. It was never a good solution in a first place, because there are risks with overflowing the cookies and other issues.

So, if you want save something temporarily in between of requests:
- Get `$requestStack` either by DI or by using `PrestaShop\PrestaShop\Adapter\SymfonyContainer::getInstance()->get('request_stack')` in legacy environment.
- Store data by using `$requestStack->getSession()->set('key', $value);`
- Get data by using `$requestStack->getSession()->get('key');`

**Comment by jolelievre** [comment]
@Hlavtox the temporary persistence you are defining is not related to the login/authorization migration Using session between requests was always possible and still is, but it's only stored for ONE request, I prefer not to confuse this part of the doc with something unrelated.

However, I'm gonna add a paragraph about persisting data with the Token, which is an alternative to using the legacy Cookie and is more in tune with this feature

**Comment by jolelievre** [comment]
@Hlavtox Actually you're right, session is persisted between multiple requests I was mixing things with flash bags that only last one request I'll follow your suggestion as it's probably safer to encourage people to use the session rather than the token attributes as we do internally In our case I think it's fine because we store data relevant for the authorization process, but it shouldn't be used for more generic purposes

**Comment by jolelievre** [comment]
@Hlavtox @kpodemski this should be good, with extra explanation on how to use the session, this page is getting a bit big but I want to finish each section first and once it's done the page will likely be reworked Some of these sections will probably be moved in a dedicated page

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
If you need to persist some custom Employee data, we don't recommend using the `Context::$cookie` anymore, instead you can use the `Session` from Symfony and update its attributes:
```

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
// You can then get/set a custom attribute, but you should always check that the session object is indeed available. Some requests do not rely on the session, and the session may never be created.
```

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
  // The session object will be automatically serialized at the end of the request, and automatically unserialized at the beginning of the next requests
```

File: modules/core-updates/9.0.md
Comment (kpodemski) [inline modules/core-updates/9.0.md]:
```suggestion
The Symfony session has other interesting features (like flash bags for unique usage data). You can read more about its usage in the <a href="https://symfony.com/doc/6.4/session.html" target="_blank">Session documentation</a>.
```

**Comment by kpodemski** [comment]
@jolelievre nice 👍🏻

**Review by Hlavtox — APPROVED** [review]
Good job, you're the boss 🎉

**Comment by jolelievre** [comment]
Last changes just pushed, @kpodemski just waiting for your approval and I think we can merge this one
I will then rebase the next one https://github.com/PrestaShop/docs/pull/1827

**Review by kpodemski — APPROVED** [review]
You have my approval @jolelievre :D

**Pull request merged**
